src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,190 @@
+
+            var issues = new List<PackagingLogMessage>();
+
+            if (noExactMatchesFromFile.Count != 0 || noExactMatchesFromNuspec.Count != 0)
+            {
+                issues.Add(PackagingLogMessage.CreateWarning(string.Join("\n", MessageFormat, noExactMatchString),
+                    NuGetLogCode.NU5128));
+            }
+
+            if (compatNotExactMatches.Count != 0)
+            {
+                issues.Add(PackagingLogMessage.CreateWarning(string.Join("\n", CompatMatchFoundWarningMessageFormat, compatMatchString),
+                    NuGetLogCode.NU5130));
+            }
+
+            if (issues.Count != 0)
+            {
+                return issues;
+            }
+
+            return Array.Empty<PackagingLogMessage>();
+        }
+
+        internal (HashSet<NuGetFramework>, HashSet<NuGetFramework>, HashSet<NuGetFramework>) CatagoriseTFMs(IEnumerable<string> files, IEnumerable<NuGetFramework> tfmsFromNuspec)
+        {
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            Func<object, object, bool> isCompatible = managedCodeConventions.Properties["tfm"].CompatibilityTest;
+            var collection = new ContentItemCollection();
+            collection.Load(files);
+
+            var libItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileLibAssemblies);
+            var refItems = ContentExtractor.GetContentForPattern(collection, managedCodeConventions.Patterns.CompileRefAssemblies);
+
+            var tfmsFromFilesSet = new HashSet<NuGetFramework>();
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(libItems));
+            tfmsFromFilesSet.AddRange(ContentExtractor.GetGroupFrameworks(refItems));
+
+            var tfmsFromFiles = tfmsFromFilesSet.Where(t => t != Net00).ToList();
+
+            var noExactMatchesFromFile = new HashSet<NuGetFramework>();
+            var noExactMatchesFromNuspec = new HashSet<NuGetFramework>();
+            var compatNotExactMatches = new HashSet<NuGetFramework>();
+
+            foreach (var fileTFM in tfmsFromFiles)
+            {
+                if (!tfmsFromNuspec.Contains(fileTFM))
+                {
+                    noExactMatchesFromFile.Add(fileTFM);
+                }
+            }
+
+            foreach (var nuspecTFM in tfmsFromNuspec)
+            {
+                if (!tfmsFromFiles.Contains(nuspecTFM))
+                {
+                    noExactMatchesFromNuspec.Add(nuspecTFM);
+                }
+            }
+
+            foreach (var fileTFM in noExactMatchesFromFile)
+            {
+                foreach (var nuspecTFM in noExactMatchesFromNuspec)
+                {
+                    if (isCompatible(fileTFM, nuspecTFM))
+                    {
+                        compatNotExactMatches.Add(fileTFM);
```

Comment:
After the first time `fileTFM` is found to be compatible with a `nuspecTFM`, there's no benefit to checking against other `nuspecTFMs`. Add a `break;` to exit out of the inner `foreach`.